File: Task/Largest-number-divisible-by-its-digits/PascalABC.NET/largest-number-divisible-by-its-digits.pas
function digits(n: int64; base: byte): List<byte>;
begin
  result := new List<byte>;
  repeat
    result.Add(byte(n mod base));
    n := n div base;
  until n = 0;
end;

function isLynchBell(num: int64; base: byte) := digits(num, base).All(x -> num mod x = 0);

begin
  for var n := 9876432 downto 1 do
  begin
    var dignum := digits(n, 10);
    if (dignum.Count = dignum.ToSet.Count) and (not dignum.Contains(0)) then
      if islynchbell(n, 10) then
      begin
        Println('Largest decimal number is:', n);
        break;
      end;
  end;

  var magic := 15 * 14 * 13 * 12 * 11;
  var n := $fedcba987654321 div magic * magic;
  while n > 0 do
  begin
    var dignum := digits(n, 16);
    if (dignum.Count = dignum.ToSet.Count) and (not dignum.Contains(0)) then
      if islynchbell(n, 16) then
      begin
        Println('Largest hex number is:', n.ToString('x'));
        break;
      end;
    n -= magic;
  end;
end.

File: Task/Largest-number-divisible-by-its-digits/JavaScript/largest-number-divisible-by-its-digits.js
function is_lynchbell(n) {
  const digits = [...String(n)].map(Number);
  if (digits.includes(0) || digits.includes(5)) {
    return false;
  }
  for (const d of digits) {
    if (digits.filter((x) => x == d).length > 1) {
      return false;
    } else if (!Number.isInteger(n / d)) {
      return false;
    }
  }
  return true;
}

for (let i = 9876432; i > 1; i--) {
  if (is_lynchbell(i)) {
    console.log(i);
    break;
  }
}
